include/drivers/video.h

```diff
@@ -0,0 +1,566 @@
+/**
+ * @file
+ *
+ * @brief Public APIs for Video.
+ */
+
+/*
+ * Copyright (c) 2019 Linaro Limited.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_VIDEO_H_
+#define ZEPHYR_INCLUDE_VIDEO_H_
+
+/**
+ * @brief Video Interface
+ * @defgroup video_interface Video Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <device.h>
+#include <stddef.h>
+#include <zephyr.h>
+
+#include <zephyr/types.h>
+
+#include <drivers/video-controls.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/**
+ * @brief video format structure
+ *
+ * Used to configure frame format.
+ *
+ * @param pixelformat is the fourcc pixel format value.
+ * @param width is the frame width in pixels.
+ * @param height is the frame height in pixels.
+ * @param pitch is the line stride, the number of bytes that needs to be added
+ *    to the address in the first pixel of a row in order to go to the address
+ *    of the first pixel of the next row (>=width).
+ */
+struct video_format {
+	u32_t pixelformat;
+	u32_t width;
+	u32_t height;
+	u32_t pitch;
+};
+
+/**
+ * @brief video format capability
+ *
+ * Used to describe a video endpoint format capability.
+ *
+ * @param pixelformat is a list of suported pixel formats (0 terminated).
+ * @param width_min is the minimum supported frame width.
+ * @param width_max is the maximum supported frame width.
+ * @param height_min is the minimum supported frame width.
+ * @param height_max is the maximum supported frame width.
+ * @param width_step is the width step size.
+ * @param height_step is the height step size.
+ */
+struct video_format_cap {
+	u32_t pixelformat;
+	u32_t width_min;
+	u32_t width_max;
+	u32_t height_min;
+	u32_t height_max;
+	u16_t width_step;
+	u16_t height_step;
+};
+
+/**
+ * @brief video capabilities
+ *
+ * Used to describe video endpoint capabilities.
+ *
+ * @param format_caps is a list of video format capabilities (zero terminated).
+ * @param min_vbuf_count is the minimal count of video buffers to enqueue
+ *    before being able to start the stream.
+ */
+struct video_caps {
+	const struct video_format_cap *format_caps;
+	u8_t min_vbuf_count;
+};
+
+/**
+ * @brief video buffer structure
+ *
+ * Represent a video frame.
+ *
+ * @param driver_data is a pointer to driver specific data.
+ * @param buffer is a pointer to the start of the buffer.
+ * @param size is the size in bytes of the buffer.
+ * @param bytesused is the number of bytes occupied by the valid data in
+ *        the buffer.
+ * @param timestamp is a time reference in milliseconds at which the last data
+ *        byte was actually received for input endpoints or to be consumed for
+ *        output endpoints.
+ */
+struct video_buffer {
+	void *driver_data;
```

Comment:
Would it no be better to change this to a struct with as first member a type identifier and as second member the data pointer?

Reply:
Well, not sure why adding a type here, this is a data-agnostic pointer (private to the driver), each video driver can use its own way. E.g. in existing video drivers a fifo is used and this private pointer is used for fifo managment. It's not restricted to specific types.

Reply:
Ok my misunderstanding, I thought it was used to communicate driver specific data between endpoints.